src/scales/scale.time.js

```diff
@@ -364,42 +358,60 @@ function ticksFromTimestamps(scale, values, majorUnit) {
 	return (ilen === 0 || !majorUnit) ? ticks : setMajorTicks(scale, ticks, map, majorUnit);
 }
 
-
 function getDataTimestamps(scale) {
 	var timestamps = scale._cache.data || [];
 	var i, ilen, metas;
 
-	if (!timestamps.length) {
-		metas = scale._getMatchingVisibleMetas();
-		for (i = 0, ilen = metas.length; i < ilen; ++i) {
-			timestamps = timestamps.concat(metas[i].controller._getAllParsedValues(scale));
-		}
-		timestamps = scale._cache.data = arrayUnique(timestamps).sort(sorter);
+	if (timestamps.length) {
+		return timestamps;
 	}
-	return timestamps;
+
+	metas = scale._getMatchingVisibleMetas();
+	for (i = 0, ilen = metas.length; i < ilen; ++i) {
+		timestamps = timestamps.concat(metas[i].controller._getAllParsedValues(scale));
+	}
+
+	// We can not assume data is in order or unique - not even for single dataset
+	// It seems to be somewhat faster to do sorting first
+	return (scale._cache.data = arrayUnique(timestamps.sort(sorter)));
 }
 
 function getLabelTimestamps(scale) {
 	var timestamps = scale._cache.labels || [];
 	var i, ilen, labels;
 
-	if (!timestamps.length) {
-		labels = scale._getLabels();
-		for (i = 0, ilen = labels.length; i < ilen; ++i) {
-			timestamps.push(parse(scale, labels[i]));
-		}
-		timestamps = scale._cache.labels = arrayUnique(timestamps).sort(sorter);
+	if (timestamps.length) {
+		return timestamps;
 	}
-	return timestamps;
+
+	labels = scale._getLabels();
+	for (i = 0, ilen = labels.length; i < ilen; ++i) {
+		timestamps.push(parse(scale, labels[i]));
+	}
+
+	// We could assume labels are in order and unique - but let's not
+	return (scale._cache.labels = arrayUnique(timestamps.sort(sorter)));
```

Comment:
you don't really need the parens around this expression

Reply:
eslint complains about assignment in return statement without